src/v/cloud_topics/level_zero/write_request_scheduler/write_request_scheduler.cc

```diff
@@ -0,0 +1,341 @@
+        ssx::spawn_with_gate(_gate, [this] { return bg_data_threshold(); });
+    }
+    co_return;
+}
+
+ss::future<> write_request_scheduler::stop() {
+    _as.request_abort();
+    co_await _gate.close();
+}
+
+struct shard_info {
+    ss::shard_id shard;
+    size_t bytes;
+};
+
+size_t write_request_scheduler::shard_bytes() noexcept {
+    // Count bytes but take limits into account.
+    size_t total_bytes = 0;
+    size_t total_requests = 0;
+    _stage.process(
+      [&total_bytes, &total_requests, this](
+        const l0::write_request<>& req) noexcept
+        -> checked<l0::request_processing_result, errc> {
+          total_bytes += req.size_bytes();
+          total_requests++;
+          return total_requests < _max_cardinality()
+                     && total_bytes < _max_buffer_size()
+                   ? request_processing_result::ignore_and_continue
+                   : request_processing_result::ignore_and_stop;
+      });
+    return total_bytes;
+}
+
+ss::future<> write_request_scheduler::bg_data_threshold() {
+    vassert(
+      !_test_only_disable_data_threshold, "Data threshold is not disabled");
+    while (!_as.abort_requested()) {
+        auto req = co_await _stage.wait_until(
+          _max_buffer_size(), ss::lowres_clock::time_point::max(), &_as);
+        if (req.has_error()) {
+            if (req.error() == errc::shutting_down) {
+                vlog(cd_log.debug, "bg_data_threshold: shutting down");
+                co_return;
+            } else {
+                vlog(
+                  cd_log.error,
+                  "bg_data_threshold: error waiting for write requests: {}",
+                  req.error());
+            }
+            co_return;
+        }
+        vlog(
+          cd_log.debug,
+          "bg_data_threshold: pending bytes: {}, total bytes: "
+          "{}",
+          req.value().pending_write_bytes,
+          req.value().total_write_bytes);
+
+        // Propagate to the next stage
+        _stage.process(
+          [](const l0::write_request<>&) noexcept
+            -> checked<l0::request_processing_result, errc> {
+              return l0::request_processing_result::advance_and_continue;
+          });
+    }
+}
+
+ss::future<> write_request_scheduler::apply_time_based_fallback() {
+    // The time based fallback works in three stages:
+    // 1. Collect information about the requests from every shard.
+    // 2. Decide which shard should handle the requests.
+    // 3. Tell every shard to forward its requests to the target shard.
+    // The whole process is triggered on shard 0. The end to end latency
+    // of the process is relatively high (up to 1ms) but it's fine. The
+    // method is invoked periodically (e.g. every 500ms) so the
+    // additional latency is not significant. If we have very high tput
+    // on some shards they will be excluded from the time based fallback
+    // because they will trigger the data threshold based uploads.
```

Comment:
it's fine for now, but this seems like it will have some non-optimal properties. the biggest one being that gang-scheduling across all the cores ignores arrival time. the observation is that the latency clock starts ticking when a request arrives, and only then is a cross-core action necessary. second, in the gang-scheduled approach:

> 3. Tell every shard to forward its requests to the target shard.

constraining the consolidation to a single node may be non-optimal.

Reply:
This changes in the next PR slightly. Shard 0 collects information periodically and decides that it's time to trigger an upload. Then it invokes the target shard (the one which will be uploading the object). The target shard pulls data from other shards and uploads the object.

Reply:
I'll write a doc with the high-level description of this process. The next PR omits few optimizations including the possibility for avoiding consolidation on one shard. It's easy to tweak the code so it'd schedule more than one shard to upload. The change is very trivial. But I suspect that it might not be necessary because in combination with the data threshold uploads shards will never accumulate enough data to justify such scheduling decision.